graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,356 @@
+            requireNonNull(type, "type");
+            final String id;
+
+            switch (type) {
+                case "connection_init":
+                    if (connectionInitiated) {
+                        // Already initiated, that's an error
+                        throw new Exception("4429: Too many initialisation requests");
+                    }
+                    final Object rawPayload = eventMap.get("payload");
+                    if (rawPayload != null) {
+                        connectionCtx = toMapFromJson(rawPayload);
+                    }
+                    connectionInitiated = true;
+                    writeConnectionAck(out);
+                    break;
+                case "ping":
+                    writePong(out);
+                    break;
+                case "pong":
+                    break;
+                case "subscribe":
+                    ensureInitiated();
+                    id = toStringFromJson("id", eventMap.get("id"));
+                    requireNonNull(id, "id");
+                    final Map<String, Object> payload = toMapFromJson(eventMap.get("payload"));
+                    try {
+                        if (graphqlSubscriptions.containsKey(id)) {
+                            throw new IllegalArgumentException(
+                                    "Subscription with id " + id + " already exists");
+                        }
+                        final String operationName = toStringFromJson("operationName",
+                                                                      payload.get("operationName"));
+                        final String query = toStringFromJson("query", payload.get("query"));
+                        final Map<String, Object> variables = toMapFromJson(payload.get("variables"));
+                        final Map<String, Object> extensions = toMapFromJson(payload.get("extensions"));
+
+                        final ExecutionInput.Builder executionInput =
+                                ExecutionInput.newExecutionInput()
+                                              .graphQLContext(upgradeCtx)
+                                              .graphQLContext(
+                                                      connectionCtx)
+                                              .query(query)
+                                              .variables(variables)
+                                              .operationName(
+                                                      operationName)
+                                              .extensions(extensions);
+
+                        final CompletableFuture<ExecutionResult> future =
+                                graphqlExecutor.executeGraphql(ctx, executionInput);
+
+                        future.handle(((executionResult, throwable) -> {
+                            handleBlocking(out, id, executionResult, throwable);
+                            return null;
+                        }));
```

Comment:
It isn't guaranteed which thread will complete the future, so `graphqlSubscriptions` can be handled from multiple threads (concurrent modification exception).

```suggestion
                        future.handleAsync(((executionResult, throwable) -> {
                            handleBlocking(out, id, executionResult, throwable);
                            return null;
                        }), ctx.eventLoop());
```